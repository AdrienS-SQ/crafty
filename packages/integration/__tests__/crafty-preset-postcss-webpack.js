/* global describe, it, expect */

const fs = require("fs");
const path = require("path");

const rimraf = require("rimraf");

const testUtils = require("../utils");

it("Compiles CSS within webpack", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-postcss-webpack/compiles")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();

  expect(
    testUtils.readForSnapshot("dist/js/myBundle.min.js")
  ).toMatchSnapshot();
});

it("Fails gracefully on broken markup", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-postcss-webpack/fails")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Compiles CSS within webpack, extracts CSS ('extractCSS' boolean option)", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-postcss-webpack/extract-boolean"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle-default.min.css")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle-default.min.css.map")).toBeTruthy();

  expect(
    testUtils.readForSnapshot("dist/js/myBundle.min.js")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/myBundle-default.min.css").toString("utf8")
  ).toMatchSnapshot();
});

it("Compiles CSS within webpack, extracts CSS ('extractCSS' string option)", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-postcss-webpack/extract-string"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle-string.min.css")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle-string.min.css.map")).toBeTruthy();

  expect(
    testUtils.readForSnapshot("dist/js/myBundle.min.js")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/myBundle-string.min.css").toString("utf8")
  ).toMatchSnapshot();
});

it("Compiles CSS within webpack, extracts CSS ('extractCSS' object option)", () => {
  process.chdir(
    path.join(
      __dirname,
      "../fixtures/crafty-preset-postcss-webpack/extract-object"
    )
  );
  rimraf.sync("dist");

  const result = testUtils.run(["run", "default"]);

  expect(result).toMatchSnapshot();

  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle-object.min.css")).toBeTruthy();
  expect(fs.existsSync("dist/js/myBundle-object.min.css.map")).toBeTruthy();

  expect(
    testUtils.readForSnapshot("dist/js/myBundle.min.js")
  ).toMatchSnapshot();
  expect(
    fs.readFileSync("dist/js/myBundle-object.min.css").toString("utf8")
  ).toMatchSnapshot();
});
