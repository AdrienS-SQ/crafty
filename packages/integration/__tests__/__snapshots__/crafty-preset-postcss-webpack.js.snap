// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiles CSS within webpack 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            css: __PATH__/packages/integration/fixtures/crafty-preset-postcss-webpack/compiles/dist/css
           js: __PATH__/packages/integration/fixtures/crafty-preset-postcss-webpack/compiles/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
  size  name     module                           status
  1 kB  0        ./js/app.js + 1 modules          built
  1 kB  4        postcss-loader!./css/style.scss  built
  1 kB  5        ./css/style.scss                 built
  1 kB  6        multi ./js/app.js                built

  size  name     asset                            status
  1 kB  default  myBundle.min.js                  emitted
  1 kB  default  myBundle.min.js.map              emitted

  (3 modules hidden)

  Compiled successfully!
  Δt ____ms

[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles CSS within webpack 2`] = `
"!function(t,e){if(\\"object\\"==typeof exports&&\\"object\\"==typeof module)module.exports=e();else if(\\"function\\"==typeof define&&define.amd)define([],e);else{var n=e();for(var r in n)(\\"object\\"==typeof exports?exports:t)[r]=n[r]}}(window,function(){return function(t){var e={};function n(r){if(e[r])return e[r].exports;var o=e[r]={i:r,l:!1,exports:{}};return t[r].call(o.exports,o,o.exports,n),o.l=!0,o.exports}return n.m=t,n.c=e,n.d=function(t,e,r){n.o(t,e)||Object.defineProperty(t,e,{configurable:!1,enumerable:!0,get:r})},n.r=function(t){Object.defineProperty(t,\\"__esModule\\",{value:!0})},n.n=function(t){var e=t&&t.__esModule?function(){return t.default}:function(){return t};return n.d(e,\\"a\\",e),e},n.o=function(t,e){return Object.prototype.hasOwnProperty.call(t,e)},n.p=\\"\\",n(n.s=6)}([function(t,e,n){\\"use strict\\";n.r(e);n(5);n.d(e,\\"default\\",function(){return r});var r=function t(){!function(t,e){if(!(t instanceof e))throw new TypeError(\\"Cannot call a class as a function\\")}(this,t)}},function(t,e){t.exports=function(t){var e=\\"undefined\\"!=typeof window&&window.location;if(!e)throw new Error(\\"fixUrls requires window.location\\");if(!t||\\"string\\"!=typeof t)return t;var n=e.protocol+\\"//\\"+e.host,r=n+e.pathname.replace(/\\\\/[^\\\\/]*$/,\\"/\\");return t.replace(/url\\\\s*\\\\(((?:[^)(]|\\\\((?:[^)(]+|\\\\([^)(]*\\\\))*\\\\))*)\\\\)/gi,function(t,e){var o,i=e.trim().replace(/^\\"(.*)\\"$/,function(t,e){return e}).replace(/^'(.*)'$/,function(t,e){return e});return/^(#|data:|http:\\\\/\\\\/|https:\\\\/\\\\/|file:\\\\/\\\\/\\\\/|\\\\s*$)/i.test(i)?t:(o=0===i.indexOf(\\"//\\")?i:0===i.indexOf(\\"/\\")?n+i:r+i.replace(/^\\\\.\\\\//,\\"\\"),\\"url(\\"+JSON.stringify(o)+\\")\\")})}},function(t,e,n){var r,o,i={},s=(r=function(){return window&&document&&document.all&&!window.atob},function(){return void 0===o&&(o=r.apply(this,arguments)),o}),a=function(t){var e={};return function(t){if(\\"function\\"==typeof t)return t();if(void 0===e[t]){var n=function(t){return document.querySelector(t)}.call(this,t);if(window.HTMLIFrameElement&&n instanceof window.HTMLIFrameElement)try{n=n.contentDocument.head}catch(t){n=null}e[t]=n}return e[t]}}(),f=null,u=0,c=[],l=n(1);function p(t,e){for(var n=0;n<t.length;n++){var r=t[n],o=i[r.id];if(o){o.refs++;for(var s=0;s<o.parts.length;s++)o.parts[s](r.parts[s]);for(;s<r.parts.length;s++)o.parts.push(m(r.parts[s],e))}else{var a=[];for(s=0;s<r.parts.length;s++)a.push(m(r.parts[s],e));i[r.id]={id:r.id,refs:1,parts:a}}}}function d(t,e){for(var n=[],r={},o=0;o<t.length;o++){var i=t[o],s=e.base?i[0]+e.base:i[0],a={css:i[1],media:i[2],sourceMap:i[3]};r[s]?r[s].parts.push(a):n.push(r[s]={id:s,parts:[a]})}return n}function h(t,e){var n=a(t.insertInto);if(!n)throw new Error(\\"Couldn't find a style target. This probably means that the value for the 'insertInto' parameter is invalid.\\");var r=c[c.length-1];if(\\"top\\"===t.insertAt)r?r.nextSibling?n.insertBefore(e,r.nextSibling):n.appendChild(e):n.insertBefore(e,n.firstChild),c.push(e);else if(\\"bottom\\"===t.insertAt)n.appendChild(e);else{if(\\"object\\"!=typeof t.insertAt||!t.insertAt.before)throw new Error(\\"[Style Loader]\\\\n\\\\n Invalid value for parameter 'insertAt' ('options.insertAt') found.\\\\n Must be 'top', 'bottom', or Object.\\\\n (https://github.com/webpack-contrib/style-loader#insertat)\\\\n\\");var o=a(t.insertInto+\\" \\"+t.insertAt.before);n.insertBefore(e,o)}}function v(t){if(null===t.parentNode)return!1;t.parentNode.removeChild(t);var e=c.indexOf(t);e>=0&&c.splice(e,1)}function b(t){var e=document.createElement(\\"style\\");return void 0===t.attrs.type&&(t.attrs.type=\\"text/css\\"),y(e,t.attrs),h(t,e),e}function y(t,e){Object.keys(e).forEach(function(n){t.setAttribute(n,e[n])})}function m(t,e){var n,r,o,i;if(e.transform&&t.css){if(!(i=e.transform(t.css)))return function(){};t.css=i}if(e.singleton){var s=u++;n=f||(f=b(e)),r=x.bind(null,n,s,!1),o=x.bind(null,n,s,!0)}else t.sourceMap&&\\"function\\"==typeof URL&&\\"function\\"==typeof URL.createObjectURL&&\\"function\\"==typeof URL.revokeObjectURL&&\\"function\\"==typeof Blob&&\\"function\\"==typeof btoa?(n=function(t){var e=document.createElement(\\"link\\");return void 0===t.attrs.type&&(t.attrs.type=\\"text/css\\"),t.attrs.rel=\\"stylesheet\\",y(e,t.attrs),h(t,e),e}(e),r=function(t,e,n){var r=n.css,o=n.sourceMap,i=void 0===e.convertToAbsoluteUrls&&o;(e.convertToAbsoluteUrls||i)&&(r=l(r));o&&(r+=\\"\\\\n/*# sourceMappingURL=data:application/json;base64,\\"+btoa(unescape(encodeURIComponent(JSON.stringify(o))))+\\" */\\");var s=new Blob([r],{type:\\"text/css\\"}),a=t.href;t.href=URL.createObjectURL(s),a&&URL.revokeObjectURL(a)}.bind(null,n,e),o=function(){v(n),n.href&&URL.revokeObjectURL(n.href)}):(n=b(e),r=function(t,e){var n=e.css,r=e.media;r&&t.setAttribute(\\"media\\",r);if(t.styleSheet)t.styleSheet.cssText=n;else{for(;t.firstChild;)t.removeChild(t.firstChild);t.appendChild(document.createTextNode(n))}}.bind(null,n),o=function(){v(n)});return r(t),function(e){if(e){if(e.css===t.css&&e.media===t.media&&e.sourceMap===t.sourceMap)return;r(t=e)}else o()}}t.exports=function(t,e){if(\\"undefined\\"!=typeof DEBUG&&DEBUG&&\\"object\\"!=typeof document)throw new Error(\\"The style-loader cannot be used in a non-browser environment\\");(e=e||{}).attrs=\\"object\\"==typeof e.attrs?e.attrs:{},e.singleton||\\"boolean\\"==typeof e.singleton||(e.singleton=s()),e.insertInto||(e.insertInto=\\"head\\"),e.insertAt||(e.insertAt=\\"bottom\\");var n=d(t,e);return p(n,e),function(t){for(var r=[],o=0;o<n.length;o++){var s=n[o];(a=i[s.id]).refs--,r.push(a)}t&&p(d(t,e),e);for(o=0;o<r.length;o++){var a;if(0===(a=r[o]).refs){for(var f=0;f<a.parts.length;f++)a.parts[f]();delete i[a.id]}}}};var g,w=(g=[],function(t,e){return g[t]=e,g.filter(Boolean).join(\\"\\\\n\\")});function x(t,e,n,r){var o=n?\\"\\":r.css;if(t.styleSheet)t.styleSheet.cssText=w(e,o);else{var i=document.createTextNode(o),s=t.childNodes;s[e]&&t.removeChild(s[e]),s.length?t.insertBefore(i,s[e]):t.appendChild(i)}}},function(t,e){t.exports=function(t){var e=[];return e.toString=function(){return this.map(function(e){var n=function(t,e){var n=t[1]||\\"\\",r=t[3];if(!r)return n;if(e&&\\"function\\"==typeof btoa){var o=(s=r,\\"/*# sourceMappingURL=data:application/json;charset=utf-8;base64,\\"+btoa(unescape(encodeURIComponent(JSON.stringify(s))))+\\" */\\"),i=r.sources.map(function(t){return\\"/*# sourceURL=\\"+r.sourceRoot+t+\\" */\\"});return[n].concat(i).concat([o]).join(\\"\\\\n\\")}var s;return[n].join(\\"\\\\n\\")}(e,t);return e[2]?\\"@media \\"+e[2]+\\"{\\"+n+\\"}\\":n}).join(\\"\\")},e.i=function(t,n){\\"string\\"==typeof t&&(t=[[null,t,\\"\\"]]);for(var r={},o=0;o<this.length;o++){var i=this[o][0];\\"number\\"==typeof i&&(r[i]=!0)}for(o=0;o<t.length;o++){var s=t[o];\\"number\\"==typeof s[0]&&r[s[0]]||(n&&!s[2]?s[2]=n:n&&(s[2]=\\"(\\"+s[2]+\\") and (\\"+n+\\")\\"),e.push(s))}},e}},function(t,e,n){(t.exports=n(3)(!1)).push([t.i,\\"a{color:#00f}.Child{background:#000}.Parent{background:#fff}.Parent--before{color:#333}.Parent--after{color:#eee}.Button{color:orange}.Menu{width:800px}.MenuLink{color:red;background:#056ef0;width:200px}\\",\\"\\"])},function(t,e,n){var r=n(4);\\"string\\"==typeof r&&(r=[[t.i,r,\\"\\"]]);var o={hmr:!0,transform:void 0,insertInto:void 0};n(2)(r,o);r.locals&&(t.exports=r.locals)},function(t,e,n){t.exports=n(0)}])});
//# sourceMappingURL=myBundle.min.js.map"
`;

exports[`Compiles CSS within webpack, extracts CSS 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            css: __PATH__/packages/integration/fixtures/crafty-preset-postcss-webpack/extract/dist/css
           js: __PATH__/packages/integration/fixtures/crafty-preset-postcss-webpack/extract/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
  size  name     module                        status
  1 kB  0        ./js/app.js + 1 modules       built
  1 kB  2        ./css/style.scss              built
  1 kB  3        multi ./js/app.js             built

  size  name     asset                         status
  1 kB  default  myBundle-default.min.css      emitted
  1 kB  default  myBundle-default.min.css.map  emitted
  1 kB  default  myBundle.min.js               emitted
  1 kB  default  myBundle.min.js.map           emitted

  (1 module hidden)

  Compiled successfully!
  Δt ____ms

[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles CSS within webpack, extracts CSS 2`] = `
"!function(e,n){if(\\"object\\"==typeof exports&&\\"object\\"==typeof module)module.exports=n();else if(\\"function\\"==typeof define&&define.amd)define([],n);else{var t=n();for(var o in t)(\\"object\\"==typeof exports?exports:e)[o]=t[o]}}(window,function(){return function(e){var n={};function t(o){if(n[o])return n[o].exports;var r=n[o]={i:o,l:!1,exports:{}};return e[o].call(r.exports,r,r.exports,t),r.l=!0,r.exports}return t.m=e,t.c=n,t.d=function(e,n,o){t.o(e,n)||Object.defineProperty(e,n,{configurable:!1,enumerable:!0,get:o})},t.r=function(e){Object.defineProperty(e,\\"__esModule\\",{value:!0})},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,\\"a\\",n),n},t.o=function(e,n){return Object.prototype.hasOwnProperty.call(e,n)},t.p=\\"\\",t(t.s=3)}([function(e,n,t){\\"use strict\\";t.r(n);t(2);t.d(n,\\"default\\",function(){return o});var o=function e(){!function(e,n){if(!(e instanceof n))throw new TypeError(\\"Cannot call a class as a function\\")}(this,e)}},,function(e,n,t){},function(e,n,t){e.exports=t(0)}])});
//# sourceMappingURL=myBundle.min.js.map"
`;

exports[`Compiles CSS within webpack, extracts CSS 3`] = `
"a{color:#00f}.Child{background:#000}.Parent{background:#fff}.Parent--before{color:#333}.Parent--after{color:#eee}.Button{color:orange}.Menu{width:800px}.MenuLink{color:red;background:#056ef0;width:200px}

/*# sourceMappingURL=myBundle-default.min.css.map*/"
`;

exports[`Fails gracefully on broken markup 1`] = `
Object {
  "status": 1,
  "stderr": "[__:__:__] 'js_myBundle' errored after ____ ms
[__:__:__] Error: Webpack compilation failed
    ...stacktrace...
[__:__:__] 'js' errored after ____ ms
[__:__:__] 'default' errored after ____ ms
",
  "stdout": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            css: __PATH__/packages/integration/fixtures/crafty-preset-postcss-webpack/fails/dist/css
           js: __PATH__/packages/integration/fixtures/crafty-preset-postcss-webpack/fails/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
  size  name  module                           status
  1 kB  0     ./js/app.js + 1 modules          built
  1 kB  3     postcss-loader!./css/style.scss  built failed ✖
  1 kB  4     ./css/style.scss                 built
  1 kB  5     multi ./js/app.js                built

  (2 assets, 2 modules hidden)

  Failed to compile.

./css/style.scss (__PATH__/packages/crafty-preset-postcss/node_modules/css-loader??ref--4-1!__PATH__/packages/crafty-preset-postcss/node_modules/postcss-loader/lib??ref--4-2!./css/style.scss)
Module build failed: Syntax Error

(17:1) Unexpected }

  15 |         color: #eee;
  16 |     }
> 17 | }
     | ^
  18 |
  19 |

 @ ./css/style.scss 2:14-207
 @ ./js/app.js
 @ multi ./js/app.js

  Δt ____ms ✖ 1 problem (1 error, 0 warnings)

",
}
`;
